### PR TITLE
merge all changes into master

### PR DESCRIPTION
**Pull request by hakosx**
# Changes :gem:
Lots of changes

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -21,6 +21,19 @@ jobs:
...
+      uses: actions/checkout@v2
+
+    - name: Set up Python version
+      uses: actions/setup-python@v1
+      with:
+        python-version: '3.7'
+    - name: Install dependencies
+      run: pip install -r requirements.txt
+    - name: Run tests
+      run: cd src && python runTests.py --keys='${{ secrets.TEST_SETTINGS }}'
```
Comment (andreacx) [inline .github/workflows/pull-request.yml:39]: new line

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+ignore = E203, E266, E501, W503, F403, F401, E402, E711, E265
+max-line-length = 79
+max-complexity = 18
+select = B,C,E,F,W,T4,B9
```
Comment (andreacx) [inline .flake8:5]: new line

File: old.sh [hunk 3]
```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
```
Comment (andreacx) [inline old.sh:1]: delete this file

File: src/customvision/.flake8 [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+ignore = E203, E266, E501, W503, F403, F401, E402, E711, E265
+max-line-length = 79
+max-complexity = 18
+select = B,C,E,F,W,T4,B9
```
Comment (andreacx) [inline src/customvision/.flake8:5]: new line

File: src/customvision/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+config.py
+test_data/**
```
Comment (andreacx) [inline src/customvision/.gitignore:2]: new line

File: src/customvision/classifier.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        res = self.predictor.classify_image_with_no_store(
+            self.project_id, self.iteration_name, img
+        )
+        # reset the file head such that it does not affect the state of the file handle
+        img.seek(0)
+        pred_kv = dict([(i.tag_name, i.probability) for i in res.predictions])
+        best_guess = max(pred_kv, key=pred_kv.get)
+        return pred_kv, best_guess
+
+    def __chunks(self, lst, n):
```
Comment (andreacx) [inline src/customvision/classifier.py:133]: ugly name convention :/

File: src/customvision/labels.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+LABEL_LIST = [
```
Comment (andreacx) [inline src/customvision/labels.py:1]: is this file supposed to be here?

File: src/html/index.html [hunk 8]
```diff
@@ -0,0 +1,89 @@
+<!doctype html>
```
Comment (andreacx) [inline src/html/index.html:1]: is this file supposed to be here?

File: src/test/websocket_test.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+    print("prediction", r)
+    assert r[0]["name"] == "prediction"
+    assert type(r[0]["args"][0]["certainty"]) is dict
+    assert type(r[0]["args"][0]["guess"]) is str
+    assert type(r[0]["args"][0]["hasWon"]) is bool
+
+
+def test_player_correct_state_after_classify(client):
+    """TODO: implement me"""
```
Comment (andreacx) [inline src/test/websocket_test.py:78]: todo?

File: src/utilities/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+"""
+    File for declaring custom exceptions.
+"""
+
+
+class UserError(Exception):
```
Comment (andreacx) [inline src/utilities/exceptions.py:6]: docstring

File: src/utilities/keys.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        keys = json.load(configFile)
+    isProduction = False
+else:
+    raise OSError(
+        "Secret keys must either be stored as environment variables"
+        " or in file 'config.json' in src/ directory"
+    )
+
+
+class Keys:
```
Comment (andreacx) [inline src/utilities/keys.py:22]: docstring

File: src/webapp/api.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    print("===== client connected =====")
+
+
+@socketio.on("disconnect")
+def disconnect():
+    """
+        When a player disconnects from a session this function tells the
+        other player in the room who left and deletes all records in the
+        database connected to the session. Old sessions is also deleted.
+    """
```
Comment (andreacx) [inline src/webapp/api.py:50]: this is not true any more

File: src/webapp/api.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+    player_id = request.sid
+    # Players join their own room as well
+    join_room(player_id)
+    game_id = models.check_player2_in_mulitplayer(player_id)
+
+    if game_id is not None:
+        # Update mulitplayer table by inserting player_id for player_2 and
+        # change state of palyer_1 in PIG to "Ready"
+        models.update_mulitplayer(player_id, game_id)
+        models.insert_into_players(player_id, game_id, "Ready")
```
Comment (andreacx) [inline src/webapp/api.py:91]: player_1 does never get the state "ready" - not critical but a bit messy maybe.

File: src/webapp/api.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    allowed_file(image_stream)
+
+    prob_kv, best_guess = classifier.predict_image(image_stream)
+
+    player_id = request.sid
+    game_id = data["game_id"]
+    time_left = data["time_left"]
+
+    game = models.get_game(game_id)
+
```
Comment (andreacx) [inline src/webapp/api.py:175]: why new line?

File: src/webapp/api.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    has_won = correct_label == best_guess and time_left > 0
+    time_out = time_left <= 0
+
+    response = {
+        "certainty": translate_probabilities(prob_kv),
+        "guess": models.to_norwegian(best_guess),
+        "correctLabel": models.to_norwegian(correct_label),
+        "hasWon": has_won,
+    }
+    emit("prediction", response)
```
Comment (andreacx) [inline src/webapp/api.py:188]: does this work without room=game_id?

File: src/test/websocket_test.py (hunk 9 above)
Comment (hakosx) [reply src/test/websocket_test.py:78]: ja, vet ikke hvordan jeg skal gjøre dette enda

Comment (andreacx) [reply src/test/websocket_test.py:78]: 👍

File: .github/workflows/pull-request.yml (hunk 1 above)
Comment (hakosx) [reply .github/workflows/pull-request.yml:39]: done

File: .flake8 (hunk 2 above)
Comment (hakosx) [reply .flake8:5]: done

File: src/customvision/.flake8 (hunk 4 above)
Comment (hakosx) [reply src/customvision/.flake8:5]: done

File: src/customvision/.gitignore (hunk 5 above)
Comment (hakosx) [reply src/customvision/.gitignore:2]: done

File: src/customvision/classifier.py (hunk 6 above)
Comment (hakosx) [reply src/customvision/classifier.py:133]: disagree

File: src/customvision/labels.py (hunk 7 above)
Comment (hakosx) [reply src/customvision/labels.py:1]: done

File: src/html/index.html (hunk 8 above)
Comment (hakosx) [reply src/html/index.html:1]: yes

File: src/utilities/exceptions.py (hunk 10 above)
Comment (hakosx) [reply src/utilities/exceptions.py:6]: done

File: src/utilities/keys.py (hunk 11 above)
Comment (hakosx) [reply src/utilities/keys.py:22]: done

File: src/webapp/api.py (hunk 12 above)
Comment (hakosx) [reply src/webapp/api.py:50]: done

File: src/webapp/api.py (hunk 13 above)
Comment (hakosx) [reply src/webapp/api.py:91]: gidder ikke fikse dette nå

File: src/webapp/api.py (hunk 14 above)
Comment (hakosx) [reply src/webapp/api.py:175]: done

File: src/webapp/api.py (hunk 15 above)
Comment (hakosx) [reply src/webapp/api.py:188]: tror det

**Comment by hakosx** [comment]
All kritikk mot meg er uten grunnlag

**Pull request merged**